apps/extension/src/content-scripts/injected-penumbra-global.ts

```diff
@@ -65,44 +112,90 @@ const requestResponseListener = (msg: MessageEvent<unknown>) => {
           failure.cause = 'Unknown';
           break;
       }
-      request.reject(failure);
+      this._request.reject(failure);
     }
+  };
+
+  // always reject with the most important reason at time of access
+  // 1. disconnect
+  // 2. connection failure
+  // 3. request
+  private get connectionFailure() {
+    // Promise.race checks in order of the list index. so if more than one
+    // promise is settled, Promise.race responds with the earlier index
+    return Promise.race([
+      // rejects with disconnect success, rejects with disconnect failure
+      this._disconnect.promise.then(() => Promise.reject(Error('Disconnected'))),
+      // ignores connection success, rejects with connection failure
+      this._connect.promise.then(() => new Promise<never>(() => null)),
+      // rejects with previous success, rejects with previous failure
+      this._request.promise.then(() => Promise.reject(Error('Disconnected'))),
+      // this should be unreachable
+      Promise.reject(Error('Unknown failure')),
+    ]);
   }
-};
 
-// Called to request a connection to the extension.
-const postRequest = () => {
-  if (!connection.state) {
-    window.addEventListener('message', requestResponseListener);
+  private postRequest() {
+    const state = this.state();
+    if (state === true)
+      // connection is already active
+      this._request.resolve();
+    else if (state === false) {
+      // connection is already failed
+      const failure = this.connectionFailure;
+      failure.catch((u: unknown) => this._request.reject(u));
+      // a previous request may have succeeded, so return the failure directly
+      return failure;
+    } else {
+      // no request made yet. attach listener and emit
+      window.addEventListener('message', this.requestFailureListener);
+      void this._request.promise.finally(() =>
+        window.removeEventListener('message', this.requestFailureListener),
+      );
+      window.postMessage(
+        {
+          [PRAX]: PraxConnection.Request,
+        } satisfies PraxMessage<PraxConnection.Request>,
+        window.origin,
+      );
+    }
+
+    return this._request.promise;
+  }
+
+  private endConnection() {
+    // attempt actual disconnect
+    void this._connect.promise
+      .then(
+        port => {
+          port.postMessage(false);
+          port.close();
+        },
+        (e: unknown) => console.warn('Could not attempt disconnect', e),
+      )
+      .catch((e: unknown) => console.error('Disconnect failed', e));
```

Comment:
If disconnect fails, shouldn't we simply just throw here?

Reply:
the call to this promise is voided, so throwing does not affect our own behavior, but unhandled rejections may cause problems elsewhere if they are ignored.

this is a purely internal failure - if this method has been called, we won't provide the connection to any caller again anyway, so the success or failure of these actions is effectively concealed. if we conveyed this rejection to the caller, the rejection would be misleading and meaningless.

so this is used to suppress the failure which won't affect our future behavior, and also log the difference between a failure that can't attempt to disconnect, and a failure that attempted to disconnect but couldn't.